---
title: "File manipulation from the command line"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

Suppose we have a gene of interest and would like to design primers for it. In particular, we want to target the those sequences that are abundant in the environments that we're interested in.

The steps that this process would take are:

1. Identify gene of interest
2. Choose an optimum similarity threshold to cluster genes into
3. Quantify environmental abundance and presence by BLASTing against metagenomes
4. Determine which clusters to include 
5. Extract the nucleotide sequences
6. Design primers
7. Perform *in silico* qPCR to determine effectiveness and specifity.


This entire process can be done entirely on the command line. However, due to time constraints, we will only be reproducing part of this pipeline today. In particular, we will do steps 2, 4, and 5. Along the way, we will review concepts and stuff.

## Choosing an optimum similarity threshold

We have a fasta file containing all the known sequences corresponding to our gene of interest.

We have clustered it already.

```{bash}
ls
```

### A review of `grep`

The `grep` function is used to search a file (or stream) for a term of interest. For example, 

#### Exercise

## Determine which clusters to include 

