---
title: "File manipulation from the command line"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

I'm currently developing a primer design pipeline for the [GERMS lab](http://www.germslab.org/). The pipeline takes as input a `fasta` file containing the genes of interest and a directory of metagenomes. The pipeline then clusters the genes, identifies which genes are abundant in the submitted metagenomes, and designs primers for those sequences. See below for an overview of the process:

![MetaFunPrimer flowchart](data/metafunprimer.png)

Today, we will working on the *Filter* step. This step is comprised of the following:

1. Choose an optimum similarity threshold for gene clustering
2. Quantify environmental abundance and presence by BLAST the representative sequences against the metagenomes
3. Determine which clusters to include 
4. Extract the nucleotide sequences

Querying the sequences against all the metagenomes takes more time than we have today, so we will only be reproducing steps 1, 3, and 4 today.

## Choosing an optimum similarity threshold

We have a fasta file containing all the known sequences corresponding to our gene of interest.

We have clustered it already.

```{bash}
ls
```


### A review of `grep`

The `grep` function is used to search a file (or stream) for a term of interest. For example, 

#### Exercise

## Determine which clusters to include 

