<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Data visualization and analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/lumen.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 54px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h2 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h3 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h4 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h5 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h6 {
  padding-top: 59px;
  margin-top: -59px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">About</a>
</li>
<li>
  <a href="introduction.html">Introduction</a>
</li>
<li>
  <a href="lesson_1.html">Shell scripting, visualizations</a>
</li>
<li>
  <a href="lesson_2.html">Data visualization and analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Data visualization and analysis</h1>

</div>


<p>According to the first-order difference criterion, the optimal clustering similarity was 98%. However, for various reasons (mostly due to physical limitations), we ended up choosing a 95% similarity threshold.</p>
<p>After determining this, we then used the representative sequences found in <code>0.95.fa</code> and queried them against our metagenomes. Thus, for each metagenome, we had an output file containing all the relevant hits. We then used a Python script to summarize the abundance and presence of each these representative sequences across our metagenomes. This information can be found in <code>data/abundance_presence.tsv</code>.</p>
<p>Our job now is to determine which of these clusters to include. While it would be nice to design primers for all of the sequences and</p>
<p>Our goal is to reproduce the following graph:</p>
<div class="figure">
<img src="data/plot.fo_difference.2-1.png" alt="Plot of first order differences by gene cluster inclusion" />
<p class="caption">Plot of first order differences by gene cluster inclusion</p>
</div>
<p>We’ll begin by learning about visualizations. I have <a href="https://pommevilla.github.io/p3.bootcamp.r.2019/lesson_2.html">an introduction to ggplot2 here</a>, but we will begin with quick review before we start tackling this one.</p>
<div id="basic-visualizations" class="section level2">
<h2>Basic visualizations</h2>
<p>We’ll begin by loading in our required packages.</p>
<pre class="r"><code>library(ggplot2)
library(tidyverse)</code></pre>
<p>Before we tackle our goal, let’s warm up on <code>mtcars</code>.</p>
<div id="exercise-2.1" class="section level3">
<h3>Exercise 2.1</h3>
<ul>
<li>Reproduce the below graph. How can it be improved?:</li>
</ul>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
</div>
</div>
<div id="importing-and-transforming-data" class="section level2">
<h2>Importing and transforming data</h2>
<p>Let’s load in the <code>tft_stats.tsv</code> data set.</p>
<pre class="r"><code>tft_stats &lt;- read.delim(&quot;data/practice_data/tft_stats.tsv&quot;)
tft_stats</code></pre>
<pre><code>##             rank percent
## 1        Iron IV    0.00
## 2       Iron III    0.08
## 3        Iron II    6.61
## 4         Iron I   14.59
## 5      Bronze IV   17.37
## 6     Bronze III   14.47
## 7      Bronze II   10.56
## 8       Bronze I    7.74
## 9      Silver IV    6.72
## 10    Silver III    5.02
## 11     Silver II    4.02
## 12      Silver I    2.91
## 13       Gold IV    3.15
## 14      Gold III    2.18
## 15       Gold II    1.58
## 16        Gold I    0.96
## 17   Platinum IV    1.07
## 18  Platinum III    0.44
## 19   Platinum II    0.23
## 20    Platinum I    0.11
## 21    Diamond IV    0.12
## 22   Diamond III    0.04
## 23    Diamond II    0.02
## 24     Diamond I    0.01
## 25      Master -    0.00
## 26 GrandMaster -    0.00
## 27  Challenger -    0.00</code></pre>
<p>We can quickly visualize this by throwing it into <code>ggplot</code>:</p>
<pre class="r"><code>tft_stats %&gt;% 
  ggplot(aes(rank, percent)) + 
  geom_col() + 
  theme(axis.text.x = element_text(angle = 90))</code></pre>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>Note that <code>ggplot</code> will automatically reorder your data if it doesn’t think it’s sorted. We can get around this by telling <code>ggplot</code> that the data already is in the correct order (adapted from <a href="https://stackoverflow.com/a/38132711">this Stack Overflow post</a>):</p>
<pre class="r"><code>tft_stats$rank &lt;- factor(tft_stats$rank, levels = tft_stats$rank)</code></pre>
<p>Now, when we plot the data frame, the order will remain:</p>
<pre class="r"><code>tft_stats %&gt;% 
  ggplot(aes(rank, percent)) + 
  geom_col() +
  theme(axis.text.x = element_text(angle = 90)) </code></pre>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>Let’s go ahead and fix this up:</p>
<pre class="r"><code>tft_stats %&gt;% 
  ggplot(aes(rank, percent)) + 
  theme_light() +
  geom_col() +
  labs(x = &quot;Rank&quot;,
       y = &quot;Percent of player base&quot;,
       title = &quot;Distribution of players across ranks in Team Fight Tactics&quot;,
       subtitle = &quot;through August 2019; source: lolchess.gg&quot;) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) </code></pre>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>We’re going to use the <code>mutate</code> function. <code>mutate</code> lets us add columns to a data frame based on other columns. For instance, we can add a column to <code>mtcars</code> to store the ratio between <code>hp</code> and <code>wt</code>:</p>
<pre class="r"><code>mtcars %&gt;% 
  mutate(hp_to_wt = hp / wt)</code></pre>
<pre><code>##     mpg cyl  disp  hp drat    wt  qsec vs am gear carb hp_to_wt
## 1  21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4 41.98473
## 2  21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4 38.26087
## 3  22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1 40.08621
## 4  21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1 34.21462
## 5  18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2 50.87209
## 6  18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1 30.34682
## 7  14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4 68.62745
## 8  24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2 19.43574
## 9  22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2 30.15873
## 10 19.2   6 167.6 123 3.92 3.440 18.30  1  0    4    4 35.75581
## 11 17.8   6 167.6 123 3.92 3.440 18.90  1  0    4    4 35.75581
## 12 16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3 44.22604
## 13 17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3 48.25737
## 14 15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3 47.61905
## 15 10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4 39.04762
## 16 10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4 39.63864
## 17 14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4 43.03087
## 18 32.4   4  78.7  66 4.08 2.200 19.47  1  1    4    1 30.00000
## 19 30.4   4  75.7  52 4.93 1.615 18.52  1  1    4    2 32.19814
## 20 33.9   4  71.1  65 4.22 1.835 19.90  1  1    4    1 35.42234
## 21 21.5   4 120.1  97 3.70 2.465 20.01  1  0    3    1 39.35091
## 22 15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2 42.61364
## 23 15.2   8 304.0 150 3.15 3.435 17.30  0  0    3    2 43.66812
## 24 13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4 63.80208
## 25 19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2 45.51365
## 26 27.3   4  79.0  66 4.08 1.935 18.90  1  1    4    1 34.10853
## 27 26.0   4 120.3  91 4.43 2.140 16.70  0  1    5    2 42.52336
## 28 30.4   4  95.1 113 3.77 1.513 16.90  1  1    5    2 74.68605
## 29 15.8   8 351.0 264 4.22 3.170 14.50  0  1    5    4 83.28076
## 30 19.7   6 145.0 175 3.62 2.770 15.50  0  1    5    6 63.17690
## 31 15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8 93.83754
## 32 21.4   4 121.0 109 4.11 2.780 18.60  1  1    4    2 39.20863</code></pre>
<p>And we can go ahead and plot this new variable right away:</p>
<pre class="r"><code>mtcars %&gt;% 
  mutate(hp_to_wt = hp / wt) %&gt;% 
  ggplot(aes(x = hp / wt, y = qsec)) +
  geom_point() + 
  labs(x = &quot;Horsepower-to-weight ratio (hp / lb)&quot;,
       y = &quot;Quarter-mile time (seconds)&quot;,
       title = &quot;Horsepower-to-weight ratio vs. quarter-mile time&quot;) + 
  theme_light()</code></pre>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>Just like with <code>summarise</code>, we can create multiple new variables at the same time with <code>mutate</code>:</p>
<pre class="r"><code>mtcars %&gt;% 
  mutate(hp_to_wt = hp / wt,
         proj_hsec = 2 * qsec)</code></pre>
<pre><code>##     mpg cyl  disp  hp drat    wt  qsec vs am gear carb hp_to_wt proj_hsec
## 1  21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4 41.98473     32.92
## 2  21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4 38.26087     34.04
## 3  22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1 40.08621     37.22
## 4  21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1 34.21462     38.88
## 5  18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2 50.87209     34.04
## 6  18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1 30.34682     40.44
## 7  14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4 68.62745     31.68
## 8  24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2 19.43574     40.00
## 9  22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2 30.15873     45.80
## 10 19.2   6 167.6 123 3.92 3.440 18.30  1  0    4    4 35.75581     36.60
## 11 17.8   6 167.6 123 3.92 3.440 18.90  1  0    4    4 35.75581     37.80
## 12 16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3 44.22604     34.80
## 13 17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3 48.25737     35.20
## 14 15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3 47.61905     36.00
## 15 10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4 39.04762     35.96
## 16 10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4 39.63864     35.64
## 17 14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4 43.03087     34.84
## 18 32.4   4  78.7  66 4.08 2.200 19.47  1  1    4    1 30.00000     38.94
## 19 30.4   4  75.7  52 4.93 1.615 18.52  1  1    4    2 32.19814     37.04
## 20 33.9   4  71.1  65 4.22 1.835 19.90  1  1    4    1 35.42234     39.80
## 21 21.5   4 120.1  97 3.70 2.465 20.01  1  0    3    1 39.35091     40.02
## 22 15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2 42.61364     33.74
## 23 15.2   8 304.0 150 3.15 3.435 17.30  0  0    3    2 43.66812     34.60
## 24 13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4 63.80208     30.82
## 25 19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2 45.51365     34.10
## 26 27.3   4  79.0  66 4.08 1.935 18.90  1  1    4    1 34.10853     37.80
## 27 26.0   4 120.3  91 4.43 2.140 16.70  0  1    5    2 42.52336     33.40
## 28 30.4   4  95.1 113 3.77 1.513 16.90  1  1    5    2 74.68605     33.80
## 29 15.8   8 351.0 264 4.22 3.170 14.50  0  1    5    4 83.28076     29.00
## 30 19.7   6 145.0 175 3.62 2.770 15.50  0  1    5    6 63.17690     31.00
## 31 15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8 93.83754     29.20
## 32 21.4   4 121.0 109 4.11 2.780 18.60  1  1    4    2 39.20863     37.20</code></pre>
<p>Let’s go ahead and use <code>mutate</code> to add a cumulative percentage column to <code>tft_stats</code>. This is done with the <code>cumsum</code> function:</p>
<pre class="r"><code>tft_stats %&gt;% 
  mutate(cumulative_percent = cumsum(percent)) </code></pre>
<pre><code>##             rank percent cumulative_percent
## 1        Iron IV    0.00               0.00
## 2       Iron III    0.08               0.08
## 3        Iron II    6.61               6.69
## 4         Iron I   14.59              21.28
## 5      Bronze IV   17.37              38.65
## 6     Bronze III   14.47              53.12
## 7      Bronze II   10.56              63.68
## 8       Bronze I    7.74              71.42
## 9      Silver IV    6.72              78.14
## 10    Silver III    5.02              83.16
## 11     Silver II    4.02              87.18
## 12      Silver I    2.91              90.09
## 13       Gold IV    3.15              93.24
## 14      Gold III    2.18              95.42
## 15       Gold II    1.58              97.00
## 16        Gold I    0.96              97.96
## 17   Platinum IV    1.07              99.03
## 18  Platinum III    0.44              99.47
## 19   Platinum II    0.23              99.70
## 20    Platinum I    0.11              99.81
## 21    Diamond IV    0.12              99.93
## 22   Diamond III    0.04              99.97
## 23    Diamond II    0.02              99.99
## 24     Diamond I    0.01             100.00
## 25      Master -    0.00             100.00
## 26 GrandMaster -    0.00             100.00
## 27  Challenger -    0.00             100.00</code></pre>
<p>And we can now add this line to our graph as another layer:</p>
<pre class="r"><code>tft_stats %&gt;% 
  mutate(cumulative = cumsum(percent)) %&gt;% 
  ggplot(aes(x = rank, group = 1)) + 
  theme_light() + 
  geom_col(aes(y = percent)) +
  geom_line(aes(y = cumulative), color = &quot;red&quot;)</code></pre>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<div id="exercise-2.2" class="section level3">
<h3>Exercise 2.2</h3>
<p>There are several things wrong with the <code>tft_stats</code> graph.</p>
<ul>
<li>What’s wrong with the y-axis? Use <code>mutate</code> to correct this.</li>
<li>What’s wrong with the cumulative percentage line? How can we fix this? Hint: Using <a href="https://en.wikipedia.org/wiki/Feature_scaling">min-max normalization</a> (with the <code>max</code> function) will get you part of the way there. What can you multiply this rescaled number by to fit with the rest of the graph? What do you think about the clarity of this graph?</li>
<li>After fixing the line, pretty up the rest of the graph. Make the cumulative percentage line dashed.</li>
</ul>
<p>Read in <code>abundance_presence.tsv</code> with:</p>
<pre class="r"><code>presence_abundance &lt;- read.delim(&quot;data/abundance_presence.tsv&quot;)</code></pre>
<p>Recreate the graph below:</p>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>Some hints:</p>
<ul>
<li>We normalized the cumulative percentage curve in the <code>tft_stats</code> graph with the following formula:</li>
</ul>
<pre class="r"><code>tft_stats %&gt;% 
  mutate(percent = percent / 100,
         cumulative = cumsum(percent) * max(percent)) </code></pre>
<pre><code>##             rank percent cumulative
## 1        Iron IV  0.0000 0.00000000
## 2       Iron III  0.0008 0.00013896
## 3        Iron II  0.0661 0.01162053
## 4         Iron I  0.1459 0.03696336
## 5      Bronze IV  0.1737 0.06713505
## 6     Bronze III  0.1447 0.09226944
## 7      Bronze II  0.1056 0.11061216
## 8       Bronze I  0.0774 0.12405654
## 9      Silver IV  0.0672 0.13572918
## 10    Silver III  0.0502 0.14444892
## 11     Silver II  0.0402 0.15143166
## 12      Silver I  0.0291 0.15648633
## 13       Gold IV  0.0315 0.16195788
## 14      Gold III  0.0218 0.16574454
## 15       Gold II  0.0158 0.16848900
## 16        Gold I  0.0096 0.17015652
## 17   Platinum IV  0.0107 0.17201511
## 18  Platinum III  0.0044 0.17277939
## 19   Platinum II  0.0023 0.17317890
## 20    Platinum I  0.0011 0.17336997
## 21    Diamond IV  0.0012 0.17357841
## 22   Diamond III  0.0004 0.17364789
## 23    Diamond II  0.0002 0.17368263
## 24     Diamond I  0.0001 0.17370000
## 25      Master -  0.0000 0.17370000
## 26 GrandMaster -  0.0000 0.17370000
## 27  Challenger -  0.0000 0.17370000</code></pre>
<p>However, if you try to do the same thing with the <code>presence_abundance</code> data frame, you won’t get the desired result. This is because the formula above is actually shorthand for:</p>
<pre class="r"><code>tft_stats %&gt;% 
  mutate(percent = percent / 100,
         cumulative = cumsum(percent) / 1 * max(percent)) </code></pre>
<pre><code>##             rank percent cumulative
## 1        Iron IV  0.0000 0.00000000
## 2       Iron III  0.0008 0.00013896
## 3        Iron II  0.0661 0.01162053
## 4         Iron I  0.1459 0.03696336
## 5      Bronze IV  0.1737 0.06713505
## 6     Bronze III  0.1447 0.09226944
## 7      Bronze II  0.1056 0.11061216
## 8       Bronze I  0.0774 0.12405654
## 9      Silver IV  0.0672 0.13572918
## 10    Silver III  0.0502 0.14444892
## 11     Silver II  0.0402 0.15143166
## 12      Silver I  0.0291 0.15648633
## 13       Gold IV  0.0315 0.16195788
## 14      Gold III  0.0218 0.16574454
## 15       Gold II  0.0158 0.16848900
## 16        Gold I  0.0096 0.17015652
## 17   Platinum IV  0.0107 0.17201511
## 18  Platinum III  0.0044 0.17277939
## 19   Platinum II  0.0023 0.17317890
## 20    Platinum I  0.0011 0.17336997
## 21    Diamond IV  0.0012 0.17357841
## 22   Diamond III  0.0004 0.17364789
## 23    Diamond II  0.0002 0.17368263
## 24     Diamond I  0.0001 0.17370000
## 25      Master -  0.0000 0.17370000
## 26 GrandMaster -  0.0000 0.17370000
## 27  Challenger -  0.0000 0.17370000</code></pre>
<p>What does the 1 represent? How can we apply that to <code>presence_abundance</code>?</p>
</div>
</div>
<div id="choosing-an-inclusion-threshold" class="section level2">
<h2>Choosing an inclusion threshold</h2>
<p>We have most of what we need to make the desired visualization. All we need now is to replace the black line with the first-order differences, plot that, and put a vertical line to show the inclusion threshold. Just like when we calculated the first-order difference on the command line, the actual code is beyond the scope of this class. Run the following lines to add the first-order difference information to the <code>presence_abundance</code> data frame:</p>
<pre class="r"><code>presence_abundance &lt;- presence_abundance %&gt;% 
  mutate(cumulative_percent = cumsum(abundance) / sum(abundance) * max(abundance))
fo_difference &lt;- function(pos){
  left &lt;- (presence_abundance[pos, 4] - presence_abundance[1, 4]) / pos
  right &lt;- (presence_abundance[nrow(presence_abundance), 4] - presence_abundance[pos, 4]) / (nrow(presence_abundance) - pos)
  return(left - right)
}
presence_abundance$fo_diffs &lt;- sapply(1:nrow(presence_abundance), fo_difference)
presence_abundance[is.na(presence_abundance)] &lt;- 0</code></pre>
<p>Let’s go ahead and plot the first-order differences against the cumulative percentage:</p>
<pre class="r"><code>presence_abundance %&gt;% 
  mutate(fo_diffs = fo_diffs / max(fo_diffs) * max(abundance)) %&gt;% 
  ggplot(aes(x = reorder(gene, -abundance), group = 1)) +
  theme_light() + 
  geom_line(aes(y = fo_diffs)) + 
  geom_line(aes(y = cumulative_percent), color = &quot;red&quot;, linetype = &quot;dashed&quot;) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) + 
  labs(x = &quot;&quot;,
       y = &quot;Abundance&quot;,
       title = &quot;Abundance of representative genes in sample metagenomes&quot;)</code></pre>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<p>All that’s left to do is to identify the cluster that achieves the maximal first-order difference. This can be done with the following:</p>
<pre class="r"><code>(elbow &lt;- which.max(presence_abundance$fo_diffs))</code></pre>
<pre><code>## [1] 6</code></pre>
<p>Now we can add a vertical line to the plot with <code>geom_vline</code>:</p>
<pre class="r"><code>presence_abundance %&gt;% 
  mutate(fo_diffs = fo_diffs / max(fo_diffs) * max(abundance)) %&gt;% 
  ggplot(aes(x = reorder(gene, -abundance), group = 1)) +
  theme_light() + 
  geom_line(aes(y = fo_diffs)) + 
  geom_line(aes(y = cumulative_percent), color = &quot;red&quot;, linetype = &quot;dashed&quot;) + 
  geom_vline(xintercept = elbow, color = &quot;red&quot;) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) + 
  labs(x = &quot;&quot;,
       y = &quot;Abundance&quot;,
       title = &quot;Abundance of representative genes in sample metagenomes&quot;)</code></pre>
<p><img src="lesson_2_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<div id="exercise-2.3" class="section level3">
<h3>Exercise 2.3</h3>
<ul>
<li>While we have identified the optimal line, it may be the case that points which are less optimal theoretically are more desirable for other practical reasons. For instance, suppose that we are okay with inclusion thresholds within 10% of the point which maximizes the first-order difference. How would you find this neighborhood? Show it on the graph using <code>geom_vline</code>?</li>
<li>The first-order difference only took into account the abundance of each particular gene. How might we also account for the presence of gene across the metagenome?</li>
</ul>
</div>
</div>
<div id="getting-the-nucleotide-sequences-for-primer-design" class="section level2">
<h2>Getting the nucleotide sequences for primer design</h2>
<p>Now that we have decided on a cluster inclusion threshold, we can now prepare a <code>fasta</code> file for input into a primer design software. So far, we have been working with the protein sequences. <a href="https://github.com/rdpstaff/EcoFunPrimer">EcoFunPrimer</a>, the primer design software that we are preparing this for, requires a <code>fasta</code> file of nucleotide sequences. The file <code>amoa_nuc.fa</code> contains the corresponding nucleotide sequences for each of the protein sequences we have. However, the sequence names are not the same between the two files. We have the mapping between the protein and nucleotide sequences in the <code>protein_nucleotide_mapping.tsv</code>. Let’s take a look at it:</p>
<pre class="bash"><code>head data/sequence_info/protein_nucleotide_mapping.tsv</code></pre>
<pre><code>## prot_id  nuc_id
## AAA66194 L08050
## AAB08985 U51630
## AAB16816 U72670
## AAB38709 U76553
## AAB38710 U76552
## AAB48015 U15733
## AAB48534 U89833
## AAB51760 U91603
## AAB53437 U92432</code></pre>
<p>We know enough to extract the nucleotide sequence ID from the mapping file for each of the protein clusters that we’re including, and use that to search the <code>amoa_nuc.fa</code> file for the sequences.</p>
<pre class="r"><code>presence_abundance[1:6, ] </code></pre>
<pre><code>##               gene abundance presence cumulative_percent  fo_diffs
## 1 AAB38709              4100      266           1283.696 -90.84851
## 2 ABM54175              1846      254           1861.672 214.37724
## 3 AAC25057              1349      234           2284.040 270.82856
## 4 SEF68642              1249      255           2675.097 296.96094
## 5 KIO48008              1015      189           2992.890 300.83481
## 6 ADE13856              1007      114           3308.179 306.95911</code></pre>
</div>
<div id="exercise-2.4" class="section level2">
<h2>Exercise 2.4</h2>
<ul>
<li>Subset the <code>presence_abundance</code> data frame to include all the sequences up until the inclusion threshold and extract the gene name column. Read the documentation for <code>write_tsv</code> to write this out to a file called <code>abundant_genes.tsv</code>. Write this without the column names.</li>
<li>We now want to search <code>protein_nucleotide_mapping.tsv</code> for the corresponding nucleotide ID sequences. In other words, we want to grep <code>protein_nucleotide_mapping.tsv</code> with multiple search terms. Note that this is the opposite situation from what we did at the beginning of this workshop where we searched multiple files for one search term. Look through the <code>grep</code> documentation and find out which flag to use to use a file for input. Extract only the nucleotide sequence from this output.</li>
<li>Pipe this output to another <code>grep</code> command to search the <code>amoa_nuc.fa</code> file for these sequences. This will give us the sequence headers, but we also want the actual sequences. Look through the <code>grep</code> documentation (or search online) to figure out how to include a certain number of lines after a successful match. How many lines should we include afterwards? What are some considerations regarding <code>fasta</code> file content that we have to keep in mind when doing this?</li>
</ul>
<p>And that’s it! Go ahead and save this to a file called <code>abundant_genes.fa</code>.</p>
</div>

<p><br>
<br>
<strong><a href="https://github.com/pommevilla">Paul Villanueva</a></strong> 
<br>
Ph.D. Student - Bioinformatics and Computational Biology<br>
Iowa State University, Ames, IA.</p>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
