<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>File manipulation from the command line</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/lumen.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 54px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h2 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h3 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h4 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h5 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h6 {
  padding-top: 59px;
  margin-top: -59px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">About</a>
</li>
<li>
  <a href="introduction.html">Introduction</a>
</li>
<li>
  <a href="lesson_1.html">Shell scripting, visualizations</a>
</li>
<li>
  <a href="lesson_2.html">Basic visualizations</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">File manipulation from the command line</h1>

</div>


<p>I’m currently developing a primer design pipeline for the <a href="http://www.germslab.org/">GERMS lab</a>. The pipeline takes as input a <code>fasta</code> file containing the genes of interest and a directory of metagenomes. The pipeline then clusters the genes, identifies which genes are abundant in the submitted metagenomes, and designs primers for those sequences. See below for an overview of the process:</p>
<div class="figure">
<img src="data/metafunprimer.png" alt="MetaFunPrimer flowchart" />
<p class="caption">MetaFunPrimer flowchart</p>
</div>
<p>Today, we will working on the <em>Filter</em> step. This step is comprised of the following:</p>
<ol style="list-style-type: decimal">
<li>Choose an optimum similarity threshold for gene clustering</li>
<li>Quantify environmental abundance and presence by BLAST the representative sequences against the metagenomes</li>
<li>Determine which clusters to include</li>
<li>Extract the nucleotide sequences</li>
</ol>
<p>Querying the sequences against all the metagenomes takes more computational resources and time than are available to us today, so we will only be reproducing steps 1, 3, and 4.</p>
<div id="choosing-an-optimum-similarity-threshold-for-gene-clustering" class="section level1">
<h1>Choosing an optimum similarity threshold for gene clustering</h1>
<p>The first step in the filtering process is to choose an optimum clustering threshold for our sequences. Sequences don’t have to be 100% similar to code for the same function, so by clustering the genes together at a lower similarity threshold may allow us to hit all of our gene targets with fewer probes.</p>
<p>To perform the clustering, we have used the <a href="http://cd-hit.org/">CD-Hit</a> program. CD-Hit accepts as input a <code>fasta</code> file and a similarity threshold and sorts the sequences into clusters. The output from this program are in <code>data/cluster_information</code> folder. They are of form <code>x.xx.fa</code> and <code>x.xx.fa.clstr</code>, where:</p>
<ul>
<li><code>x.xx</code> is the similarity threshold</li>
<li><code>.fa.clstr</code> contains the clustering information</li>
<li><code>.fa</code> contains the sequence of each cluster representative</li>
</ul>
<p>Now that we have all the clustering information from CD-Hit for our target sequences of varying thresholds, our job is to collect all of this information, use it to select an optimal similarity threshold, and collect the representative sequences for querying against our metagenomes.</p>
<p>Let’s begin with a review of <code>grep</code>.</p>
<div id="grep" class="section level2">
<h2>grep</h2>
<p>The <code>grep</code> function searches through a file for a search term. It has the following basic syntax</p>
<pre class="bash"><code>grep &lt;flags&gt; search_term file</code></pre>
<p>For example, if we wanted to find if the word “summer” was in <code>summer_days.txt</code>, we would do the following:</p>
<pre class="bash"><code>grep &quot;summer&quot; data/practice_data/summer_days.txt</code></pre>
<pre><code>## Summer days (like the summer days)
## Summer days (like the summer days)
## Summer days (like the summer days)
## Summer days (like the summer days)</code></pre>
<p>This shows us the four lines that the word “summer” was found on. Let’s see what happens when we look for a word that isn’t there:</p>
<pre class="bash"><code>grep &quot;winter&quot; data/practice_data/summer_days.txt</code></pre>
<p>Nothing pops up, as expected.</p>
<p><code>grep</code> has a ton of flags that we can pass at the command line, only a few of which we’ll cover today. For example, we can pass the <code>-v</code> flag to <strong>invert</strong> the match, returning every line on which the match wasn’t found. So, to find all the lines on which “summer” <strong>doesn’t</strong> appear, we would do:</p>
<pre class="bash"><code>grep -v &quot;summer&quot; data/practice_data/summer_days.txt</code></pre>
<pre><code>## Summer Days - Sebastian Yatra
## I don&#39;t wanna let you go
## I don&#39;t wanna let you go
## Why you wanna break my heart?
## Try to be honest
## We can make it on our own
## We can make it on our own
## You just gotta give me time
## Just stay here through August
## 
## Please, believe in me
## Like your wildest dreams
## Like your wildest dreams
## Forever free
## Sail away with me
## Don&#39;t leave
## 
## Dance on the ocean
## They be playing our song
## Hold on to meet one more time
## Dance on the ocean
## Don&#39;t kiss me goodbye
## Don&#39;t know what you think you will find
## 
## Summer days
## Summer days
## 
## Yatra, yatra
## Fue culpa del destino y te quedaste sola
## Te puso en mi camino y nos liego la hora
## Contigo a media noche quiero estar a solas
## Lo que tu quieras te haré
## Mi corazón se estalla tocandote en la playa
## Sintiendo que la arena se convierta en amor
## Y un beso es tan profundo
## Que para los segundos
## Sintiendo que el verano se convierta en amor
## 
## Dance on the ocean
## They be playing our song
## Hold on to meet one more time
## Dance on the ocean
## Don&#39;t kiss me goodbye
## Don&#39;t know what you think you will find
## 
## Summer days
## Summer days
## 
## Dance on the ocean
## They be playing our song
## Hold on to meet one more time
## Dance on the ocean
## Don&#39;t kiss me goodbye
## Don&#39;t know what you think you will find
## 
## 
## Summer days</code></pre>
<div id="exercise-1.1" class="section level3">
<h3>Exercise 1.1</h3>
<ul>
<li>With our last command, we searched for the all the lines in which the term “summer” doesn’t appear - does the output look correct? Search the <code>man</code> page (via <code>man grep</code>) and look for the flag to make the <code>grep</code> search case-insensitive.</li>
<li>Use the <code>|</code> operator to pipe the output from the above search to <code>wc</code>. How many lines was summer found on? Is there a smarter way to do this? Hint: check the <code>man</code> page.</li>
<li>Find the flag to output the line number with the matches.</li>
</ul>
<p>By default, <code>grep</code> only outputs the lines on which the match was found. This can lead to situations where using the <code>-c</code> flag to count occurrences of a term is inaccurate. For example, let’s find all the lines on which the word “sunday” appears in <code>sunday.txt</code>:</p>
<pre class="bash"><code>grep -i sunday data/practice_data/sunday.txt</code></pre>
<pre><code>## Sunday (feat. Heize, Jay Park) - GroovyRoom 
## Sunday, Sunday
## Sunday, Sunday
## Sunday, Sunday
## Sunday, the day I look forward to the most
## Sunday, Sunday
## Sunday, Sunday
## Sunday, Sunday
## My Monday to Sunday
## Sunday, Sunday
## Sunday, Sunday
## Sunday, Sunday</code></pre>
<p>Now, let’s see what <code>-c</code> says:</p>
<pre class="bash"><code>grep -ic sunday data/practice_data/sunday.txt</code></pre>
<pre><code>## 12</code></pre>
<p>We can correct this by using the <code>-o</code> flag, which tells <code>grep</code> to output each occurrence of the search term on a separate line:</p>
<pre class="bash"><code>grep -io sunday data/practice_data/sunday.txt</code></pre>
<pre><code>## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday
## Sunday</code></pre>
<p>From here, we can pipe to <code>wc</code> to get the number of occurrences:</p>
<pre class="bash"><code>grep -io sunday data/practice_data/sunday.txt | wc -l</code></pre>
<pre><code>##       21</code></pre>
</div>
<div id="exercise-1.2" class="section level3">
<h3>Exercise 1.2</h3>
<ul>
<li>How many lines does the word “Monday” appear on in <code>monday.txt</code>? How many occurrences of “Monday” are there?</li>
</ul>
</div>
</div>
<div id="working-with-fasta-files" class="section level2">
<h2>Working with <code>fasta</code> files</h2>
<p>Now that we have a bit of practice with <code>grep</code>, let’s return to the problem at hand. The <code>data/sequence_info</code> directory has the nucleotide and protein sequences for our genes of interest. Let’s take a look at them:</p>
<pre class="bash"><code>head data/sequence_info/amoa_prot.fa</code></pre>
<pre><code>## &gt;BAA92239  coded_by=987..1817,organism=Nitrosomonas sp. TK794,definition=ammonia monooxygenase a
## msifrteeilkaakmppeavhmsrlidavyfpilvvllvgtyhmhfmllagdwdfwmdwkdrqwwpvvtpivgitycsaimyylwvnyrqpfgatlcvvclligewltrywgfywwshyplnfvtpgimlpgalmldftlyltrnwlvtalvgggffglmfxpgnwpifgpthlpivvegtllsmadymghlyvrtgtpeyvrhieqgslrtfgghttviaaffaafvsmlmfavwwylgkvyctaffyvkgkrgrivqrndvtafgeegfpegik
## &gt;BAB84331  coded_by=1073..1861,organism=Nitrosomonas sp. ENI-11,definition=ammonia monooxygenase subunit A2
## mppeavhmsrlidavyfpilivllvgtyhmhfmllagdwdfwmdwkdrqwwpvvtpivgitycsaimyylwvnyrqpfgatlcvvclligewltrywgfywwshypinfvtpgimlpgalmldftmyltrnwlvtalvgggffgvlfypgnwpifgpthlpivvegtllsmadymghmyvrtgtpeyvrhieqgslrtfgghttviaaffsafvsmlmftvwwylgkvyctaffyvkgkrgrivhrndvtafgeegfpegik
## &gt;BAB84328  coded_by=1073..1861,organism=Nitrosomonas sp. ENI-11,definition=ammonia monooxygenase subunit A-1
## mppeavhmsrlidavyfpilivllvgtyhmhfmllagdwdfwmdwkdrqwwpvvtpivgitycsaimyylwvnyrqpfgatlcvvclligewltrywgfywwshypinfvtpgimlpgalmldftmyltrnwlvtalvgggffgvlfypgnwpifgpthlpivvegtllsmadymghmyvrtgtpeyvrhieqgslrtfgghttviaaffsafvsmlmftvwwylgkvyctaffyvkgkrgrivhrndvtafgeegfpegik
## &gt;SET51056  coded_by=1286..2116,organism=Nitrosomonas europaea,definition=ammonia monooxygenase subunit A
## msifrteeilkaakmppeavhmsrlidavyfpiliillvgtyhmhfmllagdwdfwmdwkdrqwwpvvtpivgitycsaimyylwvnyrqpfgatlcvvclligewltrywgfywwshypinfvtpgimlpgalmldftlyltrnwlvtalvgggffgllfypgnwpifgpthlpivvegtllsmadymghlyvrtgtpeyvrhieqgslrtfgghttviaaffsafvsmlmftvwwylgkvyctaffyvkgkrgrivhrndvtafgeegfpegik
## &gt;SDW98167  coded_by=1286..2116,organism=Nitrosomonas europaea,definition=ammonia monooxygenase subunit A
## msifrteeilkaakmppeavhmsrlidavyfpiliillvgtyhmhfmllagdwdfwmdwkdrqwwpvvtpivgitycsaimyylwvnyrqpfgatlcvvclligewltrywgfywwshypinfvtpgimlpgalmldftlyltrnwlvtalvgggffgllfypgnwpifgpthlpivvegtllsmadymghlyvrtgtpeyvrhieqgslrtfgghttviaaffsafvsmlmftvwwylgkvyctaffyvkgkrgrivhrndvtafgeegfpegik</code></pre>
<pre class="bash"><code>head data/sequence_info/amoa_nuc.fa</code></pre>
<pre><code>## &gt;AB031869  location=987..1817,organism=Nitrosomonas sp. TK794,definition=ammonia monooxygenase a
## gtgagtatatttagaacagaagagatcctgaaagcggccaagatgccgccggaagcggtccatatgtcacgcctgattgatgcggtttattttccgattctggttgttctgttggtaggtacctaccatatgcacttcatgttgttggcaggtgactgggatttctggatggactggaaagatcgtcaatggtggcctgtagtaacacctattgtaggcattacctattgctcggcaattatgtattacctgtgggtcaactaccgtcaaccatttggtgcgactctgtgcgtagtgtgtttgctgataggtgagtggctgacacgttactggggtttctactggtggtcacactatccactcaattttgtaaccccaggtatcatgctcccgggtgcattgatgttggatttcacgctgtatctgacacgtaactggttggtgactgcattggttgggggtggattctttggcctgatgtttnnnccaggtaactggccaatctttggcccgacccatctgccaatcgttgtagaaggaacactgttgtcgatggctgactacatgggtcacctgtatgttcgtacgggtacacctgagtatgttcgtcatattgaacaaggttcattacgtacctttggtggtcacaccacagttattgcggcattcttcgctgcgtttgtatccatgctgatgtttgcagtctggtggtatcttggaaaagtttactgcacagccttcttctacgttaaaggtaaaagaggacgtattgtacagcgcaatgatgttacggcatttggtgaagaagggtttccagaggggatcaaa
## &gt;AB079055  location=1073..1861,organism=Nitrosomonas sp. ENI-11,definition=ammonia monooxygenase subunit A2
## atgccaccggaagcggttcatatgtcacgtctgattgacgcggtttattttccaattctgattgttttgctggtaggtacctaccatatgcacttcatgctgctggcaggtgactgggatttctggatggactggaaagatcgtcaatggtggccggttgtgacaccgatcgtaggaattacctactgctcggcaatcatgtattacctgtgggtcaactaccgccaaccgtttggtgcaacactgtgtgtagtgtgtctgttgattggtgaatggctgacacgttactggggattctactggtggtcacactatcctatcaatttcgtaacaccgggtattatgcttccaggtgcgctgatgctggatttcacaatgtatctgacacgcaactggctggtgacggctctggttgggggtggattcttcggtgtgctgttctatccgggtaactggccgatctttggtccgacacatctgccaatcgtcgtagaaggaacactgttgtcgatggctgactacatgggccatatgtatgttcgtacgggtacacctgagtatgttcgtcatattgagcaaggttcactgcgtacctttggtggacataccacagttattgcagcgttcttctctgcattcgtatcaatgttgatgttcacagtatggtggtatcttggaaaagtttactgtacagcctttttctacgttaaaggtaaaagaggtcgtatcgtacatcgcaatgatgttaccgcattcggtgaagaaggctttccagaggggatcaaa
## &gt;AB079054  location=1073..1861,organism=Nitrosomonas sp. ENI-11,definition=ammonia monooxygenase subunit A-1
## atgccaccggaagcggttcatatgtcacgtctgattgacgcggtttattttccaattctgattgttttgctggtaggtacctaccatatgcacttcatgctgctggcaggtgactgggatttctggatggactggaaagatcgtcaatggtggccggttgtgacaccgatcgtaggaattacctactgctcggcaatcatgtattacctgtgggtcaactaccgccaaccgtttggtgcaacactgtgtgtagtgtgtctgttgattggtgaatggctgacacgttactggggattctactggtggtcacactatcctatcaatttcgtaacaccgggtattatgcttccaggtgcgctgatgctggatttcacaatgtatctgacacgcaactggctggtgacggctctggttgggggtggattcttcggtgtgctgttctatccgggtaactggccgatctttggtccgacacatctgccaatcgtcgtagaaggaacactgttgtcgatggctgactacatgggccatatgtatgttcgtacgggtacacctgagtatgttcgtcatattgagcaaggttcactgcgtacctttggtggacataccacagttattgcagcgttcttctctgcattcgtatcaatgttgatgttcacagtatggtggtatcttggaaaagtttactgtacagcctttttctacgttaaaggtaaaagaggtcgtatcgtacatcgcaatgatgttaccgcattcggtgaagaaggctttccagaggggatcaaa
## &gt;FOID01000069  location=1286..2116,organism=Nitrosomonas europaea,definition=ammonia monooxygenase subunit A
## gtgagtatatttagaacggaagaaatcctgaaagcggccaagatgccgccggaagcggttcatatgtcacgtctgattgatgcagtttattttccaattctgattattttgctggtgggtacctaccacatgcactttatgctgctggcaggtgactgggatttctggatggactggaaagatcgtcaatggtggccggttgtaacgccaatcgtggggatcacctactgttcggcaatcatgtattacttgtgggtcaactaccgccaaccgtttggtgcaacgttgtgtgtggtgtgtctgctgattggtgagtggctgacacgttactggggattctactggtggtcacactaccccatcaacttcgtaacaccgggcattatgcttccgggtgcgctgatgctggacttcacgctgtatctgacacgcaactggctggtgacggctctggttggaggtggattcttcggtctgctgttctatccgggtaactggccgatttttggaccaacccatttgccaatcgttgtagaaggcacattgctgtcgatggctgattacatgggacatctgtatgttcgtacaggtacacccgagtatgttcgtcatattgagcaaggttcactgcgtacctttggtggtcataccacagttattgcagcattcttctctgcgttcgtatcaatgttgatgttcaccgtatggtggtatcttggaaaagtttactgtacagcctttttctacgttaaaggtaaaagaggtcgtatcgtacatcgcaatgatgttaccgcattcggtgaagaaggctttccagaggggatcaaa
## &gt;FNNX01000068  location=1286..2116,organism=Nitrosomonas europaea,definition=ammonia monooxygenase subunit A
## gtgagtatatttagaacggaagaaatcctgaaagcggccaagatgccgccggaagcggttcatatgtcacgtctgattgatgcagtttattttccaattctgattattttgctggtgggtacctaccacatgcactttatgctgctggcaggtgactgggatttctggatggactggaaagatcgtcaatggtggccggttgtaacgccaatcgtggggatcacctactgttcggcaatcatgtattacttgtgggtcaactaccgccaaccgtttggtgcaacgttgtgtgtggtgtgtctgctgattggtgagtggctgacacgttactggggattctactggtggtcacactaccccatcaacttcgtaacaccgggcattatgcttccgggtgcgctgatgctggacttcacgctgtatctgacacgcaactggctggtgacggctctggttggaggtggattcttcggtctgctgttctatccgggtaactggccgatttttggaccaacccatttgccaatcgttgtagaaggcacattgctgtcgatggctgattacatgggacatctgtatgttcgtacaggtacacccgagtatgttcgtcatattgagcaaggttcactgcgtacctttggtggtcataccacagttattgcagcattcttctctgcgttcgtatcaatgttgatgttcaccgtatggtggtatcttggaaaagtttactgtacagcctttttctacgttaaaggtaaaagaggtcgtatcgtacatcgcaatgatgttaccgcattcggtgaagaaggctttccagaggggatcaaa</code></pre>
<div id="exercise-1.3" class="section level3">
<h3>Exercise 1.3</h3>
<ul>
<li>The <code>amoa_nuc.fa</code> and <code>amoa_prot.fa</code> files represent nucleotide and protein sequences of the same gene. What are some sanity checks you can do to make sure these files pass a sanity check?</li>
<li>What do you know about <code>fasta</code> files that you can use to count the number of sequences in a file?</li>
<li>Remember that our goal is to collect the number of clusters found at each of the different similarity thresholds. This information is in the <code>.fa.clstr</code> files. How are those files structured? How would you count the number of clusters in a single file? How can you expand this command to do this over all the files?</li>
<li>A large part of programming is being able to effectively search for answers and adapt them to your own needs. Modify the answer in <a href="https://stackoverflow.com/questions/34134756/tab-separated-output-from-grep">this Stack Overflow post</a> to make your output look what is printed out below, and save it to a file called <code>cluster_counts.tsv</code>:</li>
</ul>
<pre><code>## 0.80   3
## 0.81   3
## 0.82   4
## 0.83   5
## 0.84   5
## 0.85   6
## 0.86   8
## 0.87   8
## 0.88   11
## 0.89   11
## 0.90   14
## 0.91   20
## 0.92   22
## 0.93   25
## 0.94   27
## 0.95   32
## 0.96   60
## 0.97   96
## 0.98   174
## 0.99   355
## 1.00   760</code></pre>
</div>
</div>
<div id="sorting-files" class="section level2">
<h2>Sorting files</h2>
<p>So we now have the clustering information for the different similarity thresholds. To decide the threshold at which we’re going to cluster, we’re going to choose the point on this curve that maximizes the first-order difference (see <a href="https://pommevilla.github.io/random/elbows.html">here</a> for more info). Writing this code is outside the scope of this class, so for now, copy and paste the lines below into your terminal:</p>
<pre class="bash"><code>cat data/cluster_information/cluster_counts.tsv | awk -F &quot;\t&quot; &#39;BEGIN {OFS=&quot;\t&quot;} NR==1{max==2} {left = (max - $2) / NR; NR==20 ? right = 0: right = (-$2) / (20 - NR) ; print $1 OFS $2 OFS left - right}&#39; &gt; temp</code></pre>
<pre class="bash"><code>mv temp cluster_counts.tsv</code></pre>
<p>What the code above did was calculate the first-order differences at each point along the similarity curve, saved it to a temporary file <code>temp</code>, and overwrote the original <code>cluster_counts.tsv</code>.</p>
<p>Before we go any further, let’s review <code>cut</code> and <code>sort</code>.</p>
<p>The <code>cut</code> command returns the indicated columns from your file. For this example, we’re going to work with the <code>tft_stats.tsv</code> file. <code>tft_stats.tsv</code> (adapted from <a href="https://www.esportstales.com/teamfight-tactics/seasonal-rank-system-and-player-distribution">here</a>) shows the distribution of the player base across the various rankings of <a href="https://na.leagueoflegends.com/en/featured/events/teamfight-tactics">Teamfight Tactics</a>. Let’s see what it looks like:</p>
<pre class="bash"><code>cat data/practice_data/tft_stats.tsv</code></pre>
<pre><code>## Iron IV  0.00
## Iron III 0.08
## Iron II  6.61
## Iron I   14.59
## Bronze   IV  17.37
## Bronze   III 14.47
## Bronze   II  10.56
## Bronze   I   7.74
## Silver   IV  6.72
## Silver   III 5.02
## Silver   II  4.02
## Silver   I   2.91
## Gold IV  3.15
## Gold III 2.18
## Gold II  1.58
## Gold I   0.96
## Platinum IV  1.07
## Platinum III 0.44
## Platinum II  0.23
## Platinum I   0.11
## Diamond  IV  0.12
## Diamond  III 0.04
## Diamond  II  0.02
## Diamond  I   0.01
## Master   -   0.00
## GrandMaster  -   0
## Challenger   -   0</code></pre>
<p>A call of the <code>cut</code> function looks like:</p>
<pre class="bash"><code>cut -d &lt;char&gt; -f &lt;n&gt; &lt;file&gt;</code></pre>
<p>where:</p>
<ul>
<li><code>file</code> is your input file</li>
<li><code>-d</code> is the delimiter; <code>cut</code> will separate the columns of <code>file</code> based on this argument</li>
<li><code>-f</code> is the column you want</li>
</ul>
<p>By default, <code>cut</code> assumes that the file is tab-delimited. Since <code>tft_stats.tsv</code> is tab-delimited, we can go ahead and pull, say, just the first column:</p>
<pre class="bash"><code>cut -f 1 data/practice_data/tft_stats.tsv</code></pre>
<pre><code>## Iron
## Iron
## Iron
## Iron
## Bronze
## Bronze
## Bronze
## Bronze
## Silver
## Silver
## Silver
## Silver
## Gold
## Gold
## Gold
## Gold
## Platinum
## Platinum
## Platinum
## Platinum
## Diamond
## Diamond
## Diamond
## Diamond
## Master
## GrandMaster
## Challenger</code></pre>
<p>If we wanted the first two columns, we can do:</p>
<pre class="bash"><code>cut -f 1,2 data/practice_data/tft_stats.tsv</code></pre>
<pre><code>## Iron IV
## Iron III
## Iron II
## Iron I
## Bronze   IV
## Bronze   III
## Bronze   II
## Bronze   I
## Silver   IV
## Silver   III
## Silver   II
## Silver   I
## Gold IV
## Gold III
## Gold II
## Gold I
## Platinum IV
## Platinum III
## Platinum II
## Platinum I
## Diamond  IV
## Diamond  III
## Diamond  II
## Diamond  I
## Master   -
## GrandMaster  -
## Challenger   -</code></pre>
<p>and etc. Another usual function is <code>sort</code>. This function will sort a file by the indicated column.</p>
<p>A call to the <code>sort</code> function looks like:</p>
<pre class="bash"><code>cut -t &lt;char&gt; -k &lt;n&gt; &lt;file&gt;</code></pre>
<p>where:</p>
<ul>
<li><code>file</code> is your input file</li>
<li><code>-t</code> is the delimiter; <code>sort</code> will separate the columns of <code>file</code> based on this argument</li>
<li><code>-k</code> is the column you wish to sort on</li>
</ul>
<p>Let’s try sorting the file based on the third column:</p>
<pre class="bash"><code>sort -k 3 data/practice_data/tft_stats.tsv</code></pre>
<pre><code>## Challenger   -   0
## GrandMaster  -   0
## Iron IV  0.00
## Master   -   0.00
## Diamond  I   0.01
## Diamond  II  0.02
## Diamond  III 0.04
## Iron III 0.08
## Platinum I   0.11
## Diamond  IV  0.12
## Platinum II  0.23
## Platinum III 0.44
## Gold I   0.96
## Platinum IV  1.07
## Gold II  1.58
## Bronze   II  10.56
## Bronze   III 14.47
## Iron I   14.59
## Bronze   IV  17.37
## Gold III 2.18
## Silver   I   2.91
## Gold IV  3.15
## Silver   II  4.02
## Silver   III 5.02
## Iron II  6.61
## Silver   IV  6.72
## Bronze   I   7.74</code></pre>
<div id="exercise-1.4" class="section level3">
<h3>Exercise 1.4</h3>
<ul>
<li>By default, <code>sort</code> will sort the column based on string value. Look up the flag to sort based on numeric value. Look up the flag to reverse the sorting. Sort <code>tft_stats.tsv</code> based on the third column. Which division and rank had the most players in it?</li>
<li><code>ephraims_cards.tsv</code> is a list of my friend’s cards, their market value, the quantity he has of each, and the total combined. Use <code>cut</code> to access the third column and the fourth column. Use <code>sort</code> to find out which card has the most value, then use it again to find out which combination of card and quantity has the most value.</li>
<li>We now return to <code>cluster_counts.tsv</code>. Recall that we are trying to determine the optimum threshold at which to cluster our sequences, and that the criterion we’re using is the first-order difference. Which similarity threshold should we cluster our sequences at?</li>
</ul>
</div>
</div>
</div>

<p><br>
<br>
<strong><a href="https://github.com/pommevilla">Paul Villanueva</a></strong> 
<br>
Ph.D. Student - Bioinformatics and Computational Biology<br>
Iowa State University, Ames, IA.</p>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
